Gems/ROS2/Code/Source/Lidar/ROS2LidarSensorComponent.cpp

```diff
@@ -140,43 +138,73 @@ namespace ROS2
             return;
         }
 
-        const auto inverseLidarTM = entityTransform->GetWorldTM().GetInverse();
-        for (auto& point : lastScanResults.m_points)
-        {
-            point = inverseLidarTM.TransformPoint(point);
-        }
+        PublishRaycastResults(lastScanResults);
+    }
+
+    void ROS2LidarSensorComponent::PublishRaycastResults(const RaycastResult& results)
+    {
+        const bool isIntensityEnabled = m_lidarCore.m_lidarConfiguration.m_lidarSystemFeatures & LidarSystemFeatures::Intensity;
 
         auto* ros2Frame = GetEntity()->FindComponent<ROS2FrameComponent>();
         auto message = sensor_msgs::msg::PointCloud2();
         message.header.frame_id = ros2Frame->GetFrameID().data();
         message.header.stamp = ROS2Interface::Get()->GetROSTimestamp();
         message.height = 1;
-        message.width = lastScanResults.m_points.size();
-        message.point_step = sizeof(AZ::Vector3);
+        message.width = results.m_points.size();
+        message.point_step = sizeof(AZ::Vector3) + (isIntensityEnabled ? sizeof(float) : 0U);
         message.row_step = message.width * message.point_step;
 
-        AZStd::array<const char*, 3> point_field_names = { "x", "y", "z" };
-        for (int i = 0; i < point_field_names.size(); i++)
+        AZStd::array point_field_names = { "x", "y", "z", "intensity" };
+        const size_t numPointFields = isIntensityEnabled ? point_field_names.size() : point_field_names.size() - 1U;
+        for (int i = 0; i < numPointFields; i++)
         {
             sensor_msgs::msg::PointField pf;
             pf.name = point_field_names[i];
-            pf.offset = i * 4;
+            pf.offset = i * sizeof(float);
+            if (i == point_field_names.size() - 1U)
+            { // AZ::Vector3 may have padding.
+                pf.offset = sizeof(AZ::Vector3);
+            }
+
             pf.datatype = sensor_msgs::msg::PointField::FLOAT32;
             pf.count = 1;
             message.fields.push_back(pf);
         }
 
-        size_t sizeInBytes = lastScanResults.m_points.size() * sizeof(AZ::Vector3);
+        m_rosMessageData.clear();
+        const size_t sizeInBytes = results.m_points.size() * message.point_step;
         message.data.resize(sizeInBytes);
+        m_rosMessageData.reserve(sizeInBytes);
+
+        AZ_Assert(
+            !isIntensityEnabled || (results.m_points.size() == results.m_intensity.size()),
+            "Inconsistency in the size of raycast result attributes.");
+
+        const auto entityTransform = GetEntity()->FindComponent<AzFramework::TransformComponent>();
+        const auto inverseLidarTM = entityTransform->GetWorldTM().GetInverse();
+        for (size_t pointIdx = 0; pointIdx < results.m_points.size(); ++pointIdx)
+        {
+            const AZ::Vector3 globalPoint = inverseLidarTM.TransformPoint(results.m_points[pointIdx]);
+            const auto positionBytes = reinterpret_cast<const uint8_t*>(&globalPoint);
```

Comment:
Please use memcpy instead of reinterpret_cast, or reconsider point_cloud2_iterator.hpp.